### DISCUSSION TITLE
Add support for passing a queue to the run method to know when the server is ready.

### DISCUSSION THREAD

**Discussion by tchaton** [Kludex/uvicorn#1376] [Ideas]
### Checklist

- [X] There are no similar issues or pull requests for this yet.
- [X] I discussed this idea on the [community chat](https://gitter.im/encode/community) and feedback is positive.

### Is your feature related to a problem? Please describe.

I would like to know if the server is ready without hitting the endpoint. I would like to simply use a queue to know if the server is ready. 

### Describe the solution you would like.

Here is my current hack to know when the server is ready outside the main thread. The problem is the run method is changing on the uvicorn side and I can't patch it, otherwise I would be missing on latest features.

```py
class PatchServer(Server):
    def __init__(self, config, has_started_queue):
        super().__init__(config)
        self.has_started_queue = has_started_queue

    def run(self, sockets=None):
        self.config.setup_event_loop()
        loop = asyncio.get_event_loop()
        asyncio.ensure_future(self.serve(sockets=sockets))
        asyncio.ensure_future(self.check_is_started(self.has_started_queue))
        loop.run_forever()

    async def check_is_started(self, queue):
        while not self.started:
            await asyncio.sleep(1)
        queue.put(True)


# Code taken from https://github.com/encode/uvicorn/blob/master/uvicorn/main.py#L438
# TODO: Keep this updated to latest.
def uvicorn_run_fn(app, has_started_queue=None, **kwargs):
    config = Config(app, **kwargs)
    server = PatchServer(config=config, has_started_queue=has_started_queue)

    if (config.reload or config.workers > 1) and not isinstance(app, str):
        logger = logging.getLogger("uvicorn.error")
        logger.warning("You must pass the application as an import string to enable 'reload' or " "'workers'.")
        sys.exit(1)

    if config.should_reload:
        sock = config.bind_socket()
        supervisor = ChangeReload(config, target=server.run, sockets=[sock])
        supervisor.run()
    elif config.workers > 1:
        sock = config.bind_socket()
        supervisor = Multiprocess(config, target=server.run, sockets=[sock])
        supervisor.run()
    else:
        server.run()
```

### Describe alternatives you considered

I didn't consider any other.

### Additional context

None

**Comment by lovelydinosaur** [comment]
Hiya @tchaton - thanks for raising this.

I'm going to bump this into a discussion first, because I think that'll be a better starting point. (See https://github.com/encode/uvicorn/pull/1374)

Right now I'm generally trying to push for feature pause on uvicorn. I'd rather we don't consider new additions until the existing issue/PR backlog is comprehensively addressed. This'll help make sure we're focusing on stability and bug fixes.

Once we're confident that we've got the project management to a really nice sustainable point we're more likely to consider feature development.

**Comment by tchaton** [comment]
Thanks for answering @tomchristie. I actually found a cleaner solution. I am just afraid the run method changes on your side, but it seems quite stable.

```py
class PatchUvicornServer(uvicorn.Server):

    has_started_queue = None

    def run(self, sockets=None):
        self.config.setup_event_loop()
        loop = asyncio.get_event_loop()
        asyncio.ensure_future(self.serve(sockets=sockets))
        if self.has_started_queue:
            asyncio.ensure_future(self.check_is_started(self.has_started_queue))
        loop.run_forever()

    async def check_is_started(self, queue):
        while not self.started:
            await asyncio.sleep(1)
        queue.put("SERVER_HAS_STARTED")

app = ...

def start_server(..., has_started_queue=None)
   if has_started_queue:
      PatchUvicornServer.has_started_queue = has_started_queue
      # uvicorn is doing some uglyness by replacing uvicorn.main by click command.
      sys.modules["uvicorn.main"].Server = PatchUvicornServer
  uvicorn.run(app=app, host=host, port=port, log_level="error")
```

IMO, I would really like to see the run method take a server-class to enable customization.

```py
def run(server_cls):
       server_cls = server_cls or Server
       server = Server(config) 
```

**Reply by tchaton** [reply]
@[tomchristie](https://github.com/tomchristie) Any suggestions ?
